elyra/pipeline/kfp/kfp_authentication.py

```diff
@@ -0,0 +1,693 @@
+    # (See DEXLegacyAuthenticator implementation)
+    DEX_LEGACY = 'DEX (generic)'
+    # Supports DEX with LDAP authentication
+    # (See DEXLDAPAuthenticator implementation)
+    DEX_LDAP_CONNECTOR = 'DEX with LDAP'
+
+
+class AuthenticationError(Exception):
+    """
+    Indicates that an error occurred while an authentication request
+    was being processed.
+    """
+
+    def __init__(self,
+                 message: str,
+                 provider: Optional[str] = None,
+                 request_history: Optional[List[Tuple[str, requests.Response]]] = None):
+        """
+        Create a new AuthenticationError exception. The throw-er should
+        populate the request_history to allow for troubleshooting. List entry key is the (HTTP)
+        request URL, the value the response object.
+
+        :param message: a user friendly error message
+        :type message: str
+        :param provider: if the error is raised by an implementation of AbstractAuthenticator,
+         use the value of _type; optional, defaults to None
+        :type provider: Optional[str], optional
+        :param request_history: , defaults to None
+        :type request_history: Optional[List[Dict[str, requests.Response]]], optional
+        """
+        self._message = message
+        self._provider = provider
+        self._request_history = request_history
+
+    def get_request_history(self) -> Optional[List[Tuple[str, requests.Response]]]:
+        """
+        Returns the HTTP request history that led to this exception.
+
+        :return: A list of tuples, comprising HTTP URL and the response object
+        :rtype: Optional[List[Tuple[str, requests.Response]]]
+        """
+        return self._request_history
+
+    def request_history_to_string(self) -> Optional[str]:
+        """
+        Dump key HTTP request history into a string for logging purposes
+
+        :return: Formatted HTTP request history, which led to the failure.
+        :rtype: Optional[str]
+        """
+        output = None
+        for request_entry in self._request_history or []:
+            if output is None:
+                output = f'Request URL: {request_entry[0]} '\
+                         f'HTTP status code: {request_entry[1].status_code} '\
+                         f'response URL: {request_entry[1].url}'
+            else:
+                output = f'{output}\n'\
+                         f'Request URL: {request_entry[0]} '\
+                         f'HTTP status code: {request_entry[1].status_code} '\
+                         f'response URL: {request_entry[1].url}'
+        return output
+
+
+class KFPAuthenticator():
+    """
+    Use this class to authenticate with Kubeflow Pipelines. The authenticate
+    method delegates the actual authentication to an implementation of the
+    AbstractAuthenticator class.
+    """
+
+    def authenticate(self,
+                     api_endpoint: str,
+                     auth_type: str,
```

Comment:
I think we should use the enumerated type here: 
```suggestion
                     auth_type: SupportedAuthProviders,
```